### PR TITLE
feat: add structured file tools to task runner

### PR DESCRIPTION
**Pull request by robcoward**
## Summary
- Add `write_file`, `edit_file`, and `read_file` as `@function_tool` functions to the task runner alongside `execute_command`
- Add `FileMutationQueue` class with per-file `asyncio.Lock` to serialize concurrent writes to the same file path
- Add system prompt guidance instructing the agent to prefer file tools over shell commands for file I/O

## Test plan
- [x] 12 new tests in `task-runner/tests/test_file_tools.py` — all passing
- [x] All 200 task-runner tests pass (188 existing + 12 new)
- [ ] Verify CI build completes successfully
- [ ] Verify deployment on Kubernetes via ArgoCD

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: task-runner/main.py [hunk 1]
```diff
@@ -190,6 +192,15 @@ class TaskRunnerOutput(BaseModel):
     questions: list[str] = []
 
 
+FILE_TOOL_GUIDANCE = """
+
+## File Operations
+For reading, writing, and editing files, prefer the dedicated file tools
+(read_file, write_file, edit_file) over shell commands. These tools provide
+safer concurrent access and structured output. Use execute_command for
+non-file operations (installing packages, running tests, git commands, etc.).
```
[the remark this answers is not in the document]
Comment (robcoward) [reply task-runner/main.py]: Fixed in 0e77d87: Updated `execute_command` docstring to remove `cat` reference, and tightened `FILE_TOOL_GUIDANCE` wording to explicitly say "Do not use shell commands such as cat, echo, or sed for ordinary file operations."

File: task-runner/main.py [hunk 2]
```diff
@@ -418,6 +429,109 @@ def execute_command(command: str, working_directory: str = "/workspace") -> str:
...
+    Args:
+        path: The file path to write to.
+        content: The content to write to the file.
+    """
+    async with _file_mutation_queue.acquire(path):
+        p = Path(path)
+        p.parent.mkdir(parents=True, exist_ok=True)
+        p.write_text(content, encoding="utf-8")
+        byte_count = p.stat().st_size
+        return f"Wrote {byte_count} bytes to {path}"
```
[the remark this answers is not in the document]
Comment (robcoward) [reply task-runner/main.py]: Fixed in 0e77d87: Added `try/except (OSError, UnicodeError)` around the filesystem ops, returning a clear error string consistent with `execute_command`'s pattern.

[the remark this answers is not in the document]
Comment (robcoward) [reply task-runner/main.py]: Fixed in 0e77d87: Extracted sync I/O into `_write_file_sync`, `_edit_file_sync`, and `_read_file_sync` helpers, each called via `asyncio.to_thread()` so the event loop is not blocked.

File: task-runner/main.py [hunk 3]
```diff
@@ -418,6 +429,109 @@ def execute_command(command: str, working_directory: str = "/workspace") -> str:
...
+        old_text: The exact text to find (must match exactly once).
+        new_text: The replacement text.
+    """
+    p = Path(path)
+    if not p.is_file():
+        return f"Error: file not found: {path}"
+
+    async with _file_mutation_queue.acquire(path):
+        original = p.read_text(encoding="utf-8")
+        count = original.count(old_text)
```
[the remark this answers is not in the document]
Comment (robcoward) [reply task-runner/main.py]: Fixed in 0e77d87: Moved the `is_file()` check into `_edit_file_sync` which runs inside the lock, eliminating the TOCTOU race.

File: task-runner/main.py [hunk 4]
```diff
@@ -418,6 +429,109 @@ def execute_command(command: str, working_directory: str = "/workspace") -> str:
...
+        return f"Error: file not found: {path}"
+
+    lines = p.read_text(encoding="utf-8").splitlines()
+    if offset > 0:
+        lines = lines[offset:]
+    if limit > 0:
+        lines = lines[:limit]
+
+    numbered = [f"{i + offset + 1}\t{line}" for i, line in enumerate(lines)]
+    return "\n".join(numbered) if numbered else "(empty file)"
```
[the remark this answers is not in the document]
Comment (robcoward) [reply task-runner/main.py]: Fixed in 0e77d87: Added `offset = max(0, offset)` and `limit = max(0, limit)` clamping at the top of `_read_file_sync` to reject negative values.

File: task-runner/main.py [hunk 5]
```diff
@@ -418,6 +429,109 @@ def execute_command(command: str, working_directory: str = "/workspace") -> str:
...
+    @asynccontextmanager
+    async def acquire(self, path: str):
+        """Acquire a lock for the resolved absolute path."""
+        resolved = str(Path(path).resolve())
+        if resolved not in self._locks:
+            self._locks[resolved] = asyncio.Lock()
+        async with self._locks[resolved]:
+            yield
+
+
```
[the remark this answers is not in the document]
Comment (robcoward) [reply task-runner/main.py:470]: Fixed in 0e77d87: Added lock eviction after release — `if not lock.locked(): self._locks.pop(resolved, None)`. Cleans up entries as soon as no one is waiting, keeping the dict bounded without pruning complexity.

File: task-runner/main.py [hunk 6]
```diff
@@ -418,6 +433,140 @@ def execute_command(command: str, working_directory: str = "/workspace") -> str:
...
+        if resolved not in self._locks:
+            self._locks[resolved] = asyncio.Lock()
+        lock = self._locks[resolved]
+        async with lock:
+            yield
+        # Evict lock if no one else is waiting
+        if not lock.locked():
+            self._locks.pop(resolved, None)
+
+
```
[the remark this answers is not in the document]
Comment (robcoward) [reply task-runner/main.py:470]: Fixed in 1839789: Replaced `locked()`-based eviction with a refcount-based `_FileLockState` class. `users` is incremented before acquiring and decremented in a `finally` block, so the lock is only evicted when no coroutines are holding or waiting. Added a deterministic queued-waiters test to verify.

File: task-runner/main.py [hunk 7]
```diff
@@ -418,6 +433,140 @@ def execute_command(command: str, working_directory: str = "/workspace") -> str:
...
+    """Synchronous edit implementation run via asyncio.to_thread."""
+    p = Path(path)
+    if not p.is_file():
+        return f"Error: file not found: {path}"
+
+    original = p.read_text(encoding="utf-8")
+    count = original.count(old_text)
+    if count == 0:
+        return f"Error: no match found for the provided old_text in {path}"
+    if count > 1:
```
[the remark this answers is not in the document]
Comment (robcoward) [reply task-runner/main.py:511]: Won't fix: `str.count()` and `str.replace()` are consistent in their non-overlapping semantics — if count says 1, replace will do exactly one substitution at that position. This matches Claude Code's Edit tool and Pi's edit tool behavior. Overlapping matches in literal text replacement is an edge case that doesn't warrant the complexity of a custom matcher.

File: task-runner/tests/test_file_tools.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    # Launch two concurrent writers for the same path
+    await asyncio.gather(writer("A", 0.05), writer("B", 0.05))
+    # They must be serialized: A completes before B starts (or vice versa)
+    assert order[0].endswith("_start")
+    assert order[1].endswith("_end")
+    # Same writer must start and end consecutively
+    assert order[0][0] == order[1][0]  # e.g. A_start, A_end
+    assert order[2][0] == order[3][0]  # e.g. B_start, B_end
+
+
```
[the remark this answers is not in the document]
Comment (robcoward) [reply task-runner/tests/test_file_tools.py:203]: Fixed in 1839789: Added `test_mutation_queue_queued_waiters_serialize` — a deterministic barrier-based test that verifies A→B→C serialization with explicit handoff, exercising the eviction logic when waiters are queued.

File: task-runner/main.py [hunk 9]
```diff
@@ -418,6 +433,140 @@ def execute_command(command: str, working_directory: str = "/workspace") -> str:
...
+    limit = max(0, limit)
+
+    lines = p.read_text(encoding="utf-8").splitlines()
+    if offset > 0:
+        lines = lines[offset:]
+    if limit > 0:
+        lines = lines[:limit]
+
+    numbered = [f"{i + offset + 1}\t{line}" for i, line in enumerate(lines)]
+    return "\n".join(numbered) if numbered else "(empty file)"
```
[the remark this answers is not in the document]
Comment (robcoward) [reply task-runner/main.py]: Fixed in 1839789: Now distinguishes `(empty file)` (file has zero lines) from `(no lines in range)` (offset/limit produced no results from a non-empty file).

File: task-runner/main.py [hunk 10]
```diff
@@ -1097,7 +1266,7 @@ async def main():
             name="TaskRunner",
             instructions=full_instructions,
             model=env["OPENAI_MODEL"],
-            tools=[execute_command, discover_tools, submit_result],
+            tools=[execute_command, write_file, edit_file, read_file, discover_tools, submit_result],
```
[the remark this answers is not in the document]
Comment (robcoward) [reply task-runner/main.py:1289]: Good catch. Fixed in 3546d6a: Extended `_extract_file_operations()` to recognise `read_file`, `write_file`, and `edit_file` tool calls by extracting the `path` argument directly. Added `test_extract_file_operations_file_tools` test covering all three tools.
